docs/source/func.whirlwind_tour.rst

```diff
@@ -0,0 +1,197 @@
+torch.func Whirlwind Tour
+=========================
+
+What is torch.func?
+-------------------
+
+.. currentmodule:: torch.func
+
+torch.func, previously known as functorch, is a library for
+`JAX <https://github.com/google/jax>`_-like composable function transforms in
+PyTorch.
+
+- A "function transform" is a higher-order function that accepts a numerical
+  function and returns a new function that computes a different quantity.
+- torch.func has auto-differentiation transforms (``grad(f)`` returns a function
+  that computes the gradient of ``f``), a vectorization/batching transform
+  (``vmap(f)`` returns a function that computes ``f`` over batches of inputs),
+  and others.
+- These function transforms can compose with each other arbitrarily. For
+  example, composing ``vmap(grad(f))`` computes a quantity called
+  per-sample-gradients that stock PyTorch cannot efficiently compute today.
+
+Why composable function transforms?
+-----------------------------------
+There are a number of use cases that are tricky to do in PyTorch today:
+- computing per-sample-gradients (or other per-sample quantities)
+
+- running ensembles of models on a single machine
+- efficiently batching together tasks in the inner-loop of MAML
+- efficiently computing Jacobians and Hessians
+- efficiently computing batched Jacobians and Hessians
+
+Composing :func:`vmap`, :func:`grad`, :func:`vjp`, and :func:`jvp` transforms
+allows us to express the above without designing a separate subsystem for each.
+
+What are the transforms?
+------------------------
+
+:func:`grad` (gradient computation)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+``grad(func)`` is our gradient computation transform. It returns a new function
+that computes the gradients of ``func``. It assumes ``func`` returns a single-element
+Tensor and by default it computes the gradients of the output of ``func`` w.r.t.
+to the first input.
+
+.. code-block:: python
+
+    import torch
+    from torch.func import grad
+    x = torch.randn([])
+    cos_x = grad(lambda x: torch.sin(x))(x)
+    assert torch.allclose(cos_x, x.cos())
+
+    # Second-order gradients
+    neg_sin_x = grad(grad(lambda x: torch.sin(x)))(x)
+    assert torch.allclose(neg_sin_x, -x.sin())
+
+:func:`vmap` (auto-vectorization)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Note: :func:`vmap` imposes restrictions on the code that it can be used on. For more
+details, please read its docstring.
```

Comment:
nit: Since it's spelled out in more detail than in the docstring, we should link to the UX limitations here